text/0000-proc-macros.md

```diff
@@ -0,0 +1,417 @@
+making programs incomprehensible and of fragmenting the ecosystem).
+
+Therefore, I would like to remove them from the language. Alternatively, they
+could be incorporated into the new design by having another kind of macro
+function:
+
+```
+#[macro_with_ident]
+pub fn foo(&Token, TokenStream, &mut MacroContext) -> TokenStream;
+```
+
+where the first argument is the extra identifier.
+
+
+### Linking model
+
+Currently, procedural macros are dynamically linked with the compiler. This
+prevents the compiler being statically linked, which is sometimes desirable. An
+alternative architecture would have procedural macros compiled as independent
+programs and have them communicate with the compiler via IPC.
+
+This would have the advantage of allowing static linking for the compiler and
+would prevent procedural macros from crashing the main compiler process.
+However, designing a good IPC interface is complicated because there is a lot of
+data that might be exchanged between the compiler and the macro.
+
+I think we could first design the syntax, interfaces, etc. and later evolve into
+a process-separated model (if desired). However, if this is considered an
+essential feature of macro reform, then we might want to consider the interfaces
+more thoroughly with this in mind.
+
+
+### Interactions with constant evaluation
+
+Both procedural macros and constant evaluation are mechanisms for running Rust
+code at compile time. Currently, and under the proposed design, they are
+considered completely separate features. There might be some benefit in letting
+them interact.
+
+
+### Inline procedural macros
+
+It would nice to allow procedural macros to be defined in the crate in which
+they are used, as well as in separate crates (mentioned above). This complicates
+things since it breaks the invariant that a crate is designed to be used at
+either compile-time or runtime. I leave it for the future.
+
+
+### Specification of the macro definition function signatures
+
+As proposed, the signatures of functions used as macro definitions are hard-
+wired into the compiler. It would be more flexible to allow them to be specified
+by a lang-item. I'm not sure how beneficial this would be, since a change to the
+signature would require changing much of the procedural macro system. I propose
+leaving them hard-wired, unless there is a good use case for the more flexible
+approach.
+
+
+### Specifying delimiters
+
+Under this RFC, a function-like macro use may use either parentheses, braces, or
+square brackets. The choice of delimiter does not affect the semantics of the
+macro (the rules requiring braces or a semi-colon for macro uses in item position
+still apply).
+
+Which delimiter was used should be available to the macro implementation via the
+`MacroContext`. I believe this is maximally flexible - the macro implementation
+can throw an error if it doesn't like the delimiters used.
```

Comment:
I prefer this over hiding it from the macro implementor. I personally think that things like `vec!{1, 2, 3}` and `vec!(5, 6, 7)` should be forbidden or at least linted against.